{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    " <p><div class=\"lev1\"><a href=\"#Task-1.-Compiling-Ebola-Data\"><span class=\"toc-item-num\">Task 1.&nbsp;&nbsp;</span>Compiling Ebola Data</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-2.-RNA-Sequences\"><span class=\"toc-item-num\">Task 2.&nbsp;&nbsp;</span>RNA Sequences</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-3.-Class-War-in-Titanic\"><span class=\"toc-item-num\">Task 3.&nbsp;&nbsp;</span>Class War in Titanic</a></div></p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DATA_FOLDER = '../../ADA2017-Tutorials/02 - Intro to Pandas/Data/' # Use the data folder provided in Tutorial 02 - Intro to Pandas.\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 1. Compiling Ebola Data\n",
    "\n",
    "The `DATA_FOLDER/ebola` folder contains summarized reports of Ebola cases from three countries (Guinea, Liberia and Sierra Leone) during the recent outbreak of the disease in West Africa. For each country, there are daily reports that contain various information about the outbreak in several cities in each country.\n",
    "\n",
    "Use pandas to import these data files into a single `Dataframe`.\n",
    "Using this `DataFrame`, calculate for *each country*, the *daily average* per year of *new cases* and *deaths*.\n",
    "Make sure you handle all the different expressions for *new cases* and *deaths* that are used in the reports."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "On commence par analyser la stucture des fichiers : "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Date,Description,Totals,Conakry,Gueckedou,Macenta,Dabola,Kissidougou,Dinguiraye,Telimele,Boffa,Kouroussa,Siguiri,Pita,Mzerekore,Yomou,Dubreka,Forecariah,Kerouane,Coyah,Dalaba,Beyla,Kindia,Lola\n",
      "2014-09-11,New cases of suspects,9,,4,1,,,,,,,,,,,,1,3,,,,,\n",
      "\n",
      "\n",
      "Date,Variable,National,Bomi County,Bong County,Gbarpolu County,Grand Bassa,Grand Cape Mount,Grand Gedeh,Grand Kru,Lofa County,Margibi County,Maryland County,Montserrado County,Nimba County,River Gee County,RiverCess County,Sinoe County\n",
      "9/17/2014,Specimens collected,,,,,,,,,,,,,,,,\n",
      "\n",
      "\n",
      "date,variable,Kailahun,Kenema,Kono,Kambia,Koinadugu,Bombali,Tonkolili,Port Loko,Pujehun,Bo,Moyamba,Bonthe,Western area urban,Western area rural,National\n",
      "2014-08-18,population,\"465,048\",\"653,013\",\"325,003\",\"341,690\",\"335,471\",\"494,139\",\"434,937\",\"557,978\",\"335,574\",\"654,142\",\"278,119\",\"168,729\",\"1,040,888\",\"263,619\",\"6,348,350\"\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!cat '../../ADA2017-Tutorials/02 - Intro to Pandas/Data/ebola/guinea_data/2014-09-11.csv' | head -n 2 \n",
    "! echo \"\\n\"\n",
    "!cat '../../ADA2017-Tutorials/02 - Intro to Pandas/Data/ebola/liberia_data/2014-09-17-v125.csv' | head -n 2\n",
    "! echo \"\\n\"\n",
    "!cat '../../ADA2017-Tutorials/02 - Intro to Pandas/Data/ebola/sl_data/2014-08-18-v83.csv' | head -n 2\n",
    "! echo \"\\n\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Premiere dificulter chaque pays Ã  sont propre format mais Gloabalement similaire :\n",
    "Date , Description/Variable , (Toatal (seulemnt pour la Guinea)) , les donner pour chaque region ou county"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Voyon ce que contien le champs variable/Description comme colone :"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cat: ../../ADA2017-Tutorials/02- Intro to Pandas/Data/ebola/*/*: No such file or directory\r\n"
     ]
    }
   ],
   "source": [
    "cat ../../ADA2017-Tutorials/02-\\ Intro\\ to\\ Pandas/Data/ebola/*/* |  cut -d, -f2  | sort | uniq"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "\n",
    " suspected cases\"\n",
    "\"Case Fatality Rate (CFR) - \n",
    "\"Case Fatality Rate (CFR) - Confirmed & Probable Cases\"\n",
    "\"Contacts who completed 21 day \n",
    "\"Contacts who completed 21 day follow-up\"\n",
    "\"Cumulative confirmed\n",
    "\"Total Number of Confirmed Cases \n",
    "\"Total Number of Confirmed Cases of Guinean Nationality\"\n",
    "\"Total Number of Confirmed Cases of Sierra Leonean Nationality\"\n",
    "\"Total death/s in confirmed\n",
    "\"Total no. currently in Treatment \n",
    "\"Total no. currently in Treatment Units\"\n",
    "0\n",
    "123\n",
    "133\n",
    "134\n",
    "136\n",
    "137\n",
    "147\n",
    "148\n",
    "149\n",
    "151\n",
    "156\n",
    "164\n",
    "176\n",
    "180\n",
    "182\n",
    "205\n",
    "211\n",
    "230\n",
    "246\n",
    "299\n",
    "373\n",
    "67\n",
    "69\n",
    "70\n",
    "782\n",
    "8\n",
    "859\n",
    "94\n",
    "99\n",
    "Case Fatality Rate (CFR) - Confirmed & Probable Cases\n",
    "Contacts lost to follow-up\n",
    "Contacts seen\n",
    "Contacts who completed 21 day follow-up\n",
    "Cumulative (confirmed + probable + suspected)\n",
    "Cumulative (confirmed + probable + suspects)\n",
    "Cumulative CFR\n",
    "Cumulative admission/isolation\n",
    "Cumulative cases among HCW\n",
    "Cumulative deaths among HCW\n",
    "Currently under follow-up\n",
    "Description\n",
    "Fatality rate for confirmed and probables\n",
    "New Case/s (Probable)\n",
    "New Case/s (Suspected)\n",
    "New admissions\n",
    "New admits to CTE so far\n",
    "New case/s (confirmed)\n",
    "New cases of confirmed\n",
    "New cases of confirmed among health workers\n",
    "New cases of probables\n",
    "New cases of suspects\n",
    "New contacts registered so far\n",
    "New deaths registered\n",
    "New deaths registered among health workers\n",
    "New deaths registered today\n",
    "New deaths registered today (confirmed)\n",
    "New deaths registered today (probables)\n",
    "New deaths registered today (suspects)\n",
    "Newly Reported Cases in HCW\n",
    "Newly Reported deaths in HCW\n",
    "Newly reported contacts\n",
    "Newly reported deaths\n",
    "Number of confirmed cases among health workers\n",
    "Number of contacts followed today\n",
    "Number of contacts followed yesterday\n",
    "Number of contacts lost to follow up\n",
    "Number of contacts out of the track 21 days\n",
    "Number of contacts out of track\n",
    "Number of contacts to follow today\n",
    "Number of death of confirmed cases among health workers\n",
    "Number of deaths of confirmed cases among health workers\n",
    "Number of deaths of probables cases among health workers\n",
    "Number of female confirmed cases\n",
    "Number of female probables cases\n",
    "Number of female suspects cases\n",
    "Number of male confirmed cases\n",
    "Number of male probables cases\n",
    "Number of male suspects cases\n",
    "Number of patients tested\n",
    "Number of probables cases among health workers\n",
    "Number of samples collected\n",
    "Number of samples collected today\n",
    "Number of samples under test\n",
    "Number of suspects cases among health workers\n",
    "Specimens collected\n",
    "Specimens pending for testing\n",
    "Total Case/s (Probable)\n",
    "Total Case/s (Suspected)\n",
    "Total Number of Confirmed Cases of Guinean Nationality\n",
    "Total Number of Confirmed Cases of Sierra Leonean Nationality\n",
    "Total PEC center today\n",
    "Total PEC center today (confirmed)\n",
    "Total PEC center today (probables)\n",
    "Total PEC center today (suspects)\n",
    "Total case of confirmed among health workers\n",
    "Total case/s (confirmed)\n",
    "Total cases of confirmed\n",
    "Total cases of probables\n",
    "Total cases of suspects\n",
    "Total confirmed cases\n",
    "Total contacts listed\n",
    "Total contacts registered from start date\n",
    "Total death/s in confirmed cases\n",
    "Total death/s in probable cases\n",
    "Total death/s in suspected cases\n",
    "Total deaths (confirmed + probables + suspects)\n",
    "Total deaths of confirmed\n",
    "Total deaths of probables\n",
    "Total deaths of suspects\n",
    "Total deaths registered among health workers\n",
    "Total discharges\n",
    "Total new cases registered so far\n",
    "Total no. currently in Treatment Units\n",
    "Total number of admissions to CTE\n",
    "Total number of exits from CTE\n",
    "Total number of female cases\n",
    "Total number of hospitalized cases in CTE\n",
    "Total number of male cases\n",
    "Total of cured in confirmed cases in CTE\n",
    "Total of deaths in confirmed cases in CTE\n",
    "Total probable cases\n",
    "Total samples tested\n",
    "Total specimens tested\n",
    "Total suspected cases\n",
    "Total suspected non-class cases\n",
    "Variable\n",
    "cfr\n",
    "contacts_followed\n",
    "contacts_healthy\n",
    "contacts_ill\n",
    "contacts_not_seen\n",
    "cum_completed_contacts\n",
    "cum_confirmed\n",
    "cum_contacts\n",
    "cum_noncase\n",
    "cum_probable\n",
    "cum_suspected\n",
    "death_confirmed\n",
    "death_probable\n",
    "death_suspected\n",
    "etc_cum_admission\n",
    "etc_cum_deaths\n",
    "etc_cum_discharges\n",
    "etc_currently_admitted\n",
    "etc_new_admission\n",
    "etc_new_deaths\n",
    "etc_new_discharges\n",
    "negative_corpse\n",
    "new_completed_contacts\n",
    "new_confirmed\n",
    "new_contacts\n",
    "new_negative\n",
    "new_noncase\n",
    "new_positive\n",
    "new_probable\n",
    "new_samples\n",
    "new_suspected\n",
    "pending\n",
    "percent_seen\n",
    "population\n",
    "positive_corpse\n",
    "repeat_samples\n",
    "total_lab_samples\n",
    "variable"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "LEs chiffre vinne de fichier ou un saut de ligne a eu lieux comme par exemple :  liberia_data/2014-10-31-v169.csv\n",
    "trion un peu"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "\"Case Fatality Rate (CFR) - \n",
    "\"Case Fatality Rate (CFR) - Confirmed & Probable Cases\"\n",
    "\"Contacts who completed 21 day \n",
    "\"Contacts who completed 21 day follow-up\"\n",
    "\"Cumulative confirmed\n",
    "\"Total Number of Confirmed Cases \n",
    "\"Total Number of Confirmed Cases of Guinean Nationality\"\n",
    "\"Total Number of Confirmed Cases of Sierra Leonean Nationality\"\n",
    "\"Total death/s in confirmed\n",
    "\"Total no. currently in Treatment \n",
    "\"Total no. currently in Treatment Units\"\n",
    "Case Fatality Rate (CFR) - Confirmed & Probable Cases\n",
    "Contacts lost to follow-up\n",
    "Contacts seen\n",
    "Contacts who completed 21 day follow-up\n",
    "Cumulative (confirmed + probable + suspected)\n",
    "Cumulative (confirmed + probable + suspects)\n",
    "Cumulative CFR\n",
    "Cumulative admission/isolation\n",
    "Cumulative cases among HCW\n",
    "Cumulative deaths among HCW\n",
    "Currently under follow-up\n",
    "Fatality rate for confirmed and probables\n",
    "New Case/s (Probable)\n",
    "New Case/s (Suspected)\n",
    "New admissions\n",
    "New admits to CTE so far\n",
    "New case/s (confirmed)\n",
    "New cases of confirmed\n",
    "New cases of confirmed among health workers\n",
    "New cases of probables\n",
    "New cases of suspects\n",
    "New contacts registered so far\n",
    "New deaths registered\n",
    "New deaths registered among health workers\n",
    "New deaths registered today\n",
    "New deaths registered today (confirmed)\n",
    "New deaths registered today (probables)\n",
    "New deaths registered today (suspects)\n",
    "Newly Reported Cases in HCW\n",
    "Newly Reported deaths in HCW\n",
    "Newly reported contacts\n",
    "Newly reported deaths\n",
    "Number of confirmed cases among health workers\n",
    "Number of contacts followed today\n",
    "Number of contacts followed yesterday\n",
    "Number of contacts lost to follow up\n",
    "Number of contacts out of the track 21 days\n",
    "Number of contacts out of track\n",
    "Number of contacts to follow today\n",
    "Number of death of confirmed cases among health workers\n",
    "Number of deaths of confirmed cases among health workers\n",
    "Number of deaths of probables cases among health workers\n",
    "Number of female confirmed cases\n",
    "Number of female probables cases\n",
    "Number of female suspects cases\n",
    "Number of male confirmed cases\n",
    "Number of male probables cases\n",
    "Number of male suspects cases\n",
    "Number of patients tested\n",
    "Number of probables cases among health workers\n",
    "Number of samples collected\n",
    "Number of samples collected today\n",
    "Number of samples under test\n",
    "Number of suspects cases among health workers\n",
    "Specimens collected\n",
    "Specimens pending for testing\n",
    "Total Case/s (Probable)\n",
    "Total Case/s (Suspected)\n",
    "Total Number of Confirmed Cases of Guinean Nationality\n",
    "Total Number of Confirmed Cases of Sierra Leonean Nationality\n",
    "Total PEC center today\n",
    "Total PEC center today (confirmed)\n",
    "Total PEC center today (probables)\n",
    "Total PEC center today (suspects)\n",
    "Total case of confirmed among health workers\n",
    "Total case/s (confirmed)\n",
    "Total cases of confirmed\n",
    "Total cases of probables\n",
    "Total cases of suspects\n",
    "Total confirmed cases\n",
    "Total contacts listed\n",
    "Total contacts registered from start date\n",
    "Total death/s in confirmed cases\n",
    "Total death/s in probable cases\n",
    "Total death/s in suspected cases\n",
    "Total deaths (confirmed + probables + suspects)\n",
    "Total deaths of confirmed\n",
    "Total deaths of probables\n",
    "Total deaths of suspects\n",
    "Total deaths registered among health workers\n",
    "Total discharges\n",
    "Total new cases registered so far\n",
    "Total no. currently in Treatment Units\n",
    "Total number of admissions to CTE\n",
    "Total number of exits from CTE\n",
    "Total number of female cases\n",
    "Total number of hospitalized cases in CTE\n",
    "Total number of male cases\n",
    "Total of cured in confirmed cases in CTE\n",
    "Total of deaths in confirmed cases in CTE\n",
    "Total probable cases\n",
    "Total samples tested\n",
    "Total specimens tested\n",
    "Total suspected cases\n",
    "Total suspected non-class cases\n",
    "cfr\n",
    "contacts_followed\n",
    "contacts_healthy\n",
    "contacts_ill\n",
    "contacts_not_seen\n",
    "cum_completed_contacts\n",
    "cum_confirmed\n",
    "cum_contacts\n",
    "cum_noncase\n",
    "cum_probable\n",
    "cum_suspected\n",
    "death_confirmed\n",
    "death_probable\n",
    "death_suspected\n",
    "etc_cum_admission\n",
    "etc_cum_deaths\n",
    "etc_cum_discharges\n",
    "etc_currently_admitted\n",
    "etc_new_admission\n",
    "etc_new_deaths\n",
    "etc_new_discharges\n",
    "negative_corpse\n",
    "new_completed_contacts\n",
    "new_confirmed\n",
    "new_contacts\n",
    "new_negative\n",
    "new_noncase\n",
    "new_positive\n",
    "new_probable\n",
    "new_samples\n",
    "new_suspected\n",
    "pending\n",
    "percent_seen\n",
    "population\n",
    "positive_corpse\n",
    "repeat_samples\n",
    "total_lab_samples\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "On vois que beacoup de champs se recoupe il faudra les susioner lors de l'inportations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3588,)"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import glob\n",
    "#data = pd.DataFrame({'date':[], 'variable':[]})\n",
    "#data = data.set_index(['date', 'variable'])\n",
    "data = pd.DataFrame()\n",
    "test = 0\n",
    "for csv in  glob.glob(DATA_FOLDER+\"ebola/*/*.csv\"):\n",
    "    subframe =  pd.read_csv(csv)\n",
    "    subframe.columns = [x.lower() for x in subframe.columns]\n",
    "    if(subframe.columns[1]==\"description\") : \n",
    "        subframe.rename(columns={'description': 'variable'}, inplace=True)\n",
    "\n",
    "    uniquedate = subframe['date'][0]\n",
    "\n",
    "    subframe.drop('date',axis=1, inplace=True)\n",
    "    subframe = subframe.set_index(\"variable\")\n",
    "    subframe = subframe.transpose()\n",
    "    subframe['date']= uniquedate\n",
    "\n",
    "    \n",
    "    if \"guinea\" in csv :\n",
    "        subframe['country'] = \"Guinea\"\n",
    "    if \"liberia\" in csv :\n",
    "        subframe['country'] = \"Liberia\"\n",
    "    if \"sl\" in csv :\n",
    "        subframe['country'] = \"Siera Leon\"\n",
    "        \n",
    "        \n",
    "    subframe= subframe.groupby(subframe.columns, axis=1).sum() # remove duplicate colone name\n",
    "    #if(test>89):\n",
    "      #  print(subframe.columns.duplicated())\n",
    "       # pd.concat([data,subframe], axis=0, ignore_index=True)\n",
    "        #break\n",
    "        \n",
    "    if(len(data)==0):\n",
    "        data = subframe\n",
    "    else :\n",
    "        data = pd.concat([data,subframe], axis=0, ignore_index=True)\n",
    "        #data = pd.merge(data, subframe, how=\"outer\")\n",
    "        \n",
    "\n",
    "    test = test+1\n",
    "\n",
    "data['country'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 2. RNA Sequences\n",
    "\n",
    "In the `DATA_FOLDER/microbiome` subdirectory, there are 9 spreadsheets of microbiome data that was acquired from high-throughput RNA sequencing procedures, along with a 10<sup>th</sup> file that describes the content of each. \n",
    "\n",
    "Use pandas to import the first 9 spreadsheets into a single `DataFrame`.\n",
    "Then, add the metadata information from the 10<sup>th</sup> spreadsheet as columns in the combined `DataFrame`.\n",
    "Make sure that the final `DataFrame` has a unique index and all the `NaN` values have been replaced by the tag `unknown`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write your answer here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 3. Class War in Titanic\n",
    "\n",
    "Use pandas to import the data file `Data/titanic.xls`. It contains data on all the passengers that travelled on the Titanic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from IPython.core.display import HTML\n",
    "HTML(filename=DATA_FOLDER+'/titanic.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each of the following questions state clearly your assumptions and discuss your findings:\n",
    "1. Describe the *type* and the *value range* of each attribute. Indicate and transform the attributes that can be `Categorical`. \n",
    "2. Plot histograms for the *travel class*, *embarkation port*, *sex* and *age* attributes. For the latter one, use *discrete decade intervals*. \n",
    "3. Calculate the proportion of passengers by *cabin floor*. Present your results in a *pie chart*.\n",
    "4. For each *travel class*, calculate the proportion of the passengers that survived. Present your results in *pie charts*.\n",
    "5. Calculate the proportion of the passengers that survived by *travel class* and *sex*. Present your results in *a single histogram*.\n",
    "6. Create 2 equally populated *age categories* and calculate survival proportions by *age category*, *travel class* and *sex*. Present your results in a `DataFrame` with unique index."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write your answer here"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
