{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    " <p><div class=\"lev1\"><a href=\"#Task-1.-Compiling-Ebola-Data\"><span class=\"toc-item-num\">Task 1.&nbsp;&nbsp;</span>Compiling Ebola Data</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-2.-RNA-Sequences\"><span class=\"toc-item-num\">Task 2.&nbsp;&nbsp;</span>RNA Sequences</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-3.-Class-War-in-Titanic\"><span class=\"toc-item-num\">Task 3.&nbsp;&nbsp;</span>Class War in Titanic</a></div></p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import os\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DATA_FOLDER = '../../ADA2017-Tutorials/02 - Intro to pandas/Data/' # Use the data folder provided in Tutorial 02 - Intro to Pandas."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 2. RNA Sequences\n",
    "\n",
    "In the `DATA_FOLDER/microbiome` subdirectory, there are 9 spreadsheets of microbiome data that was acquired from high-throughput RNA sequencing procedures, along with a 10<sup>th</sup> file that describes the content of each. \n",
    "\n",
    "Use pandas to import the first 9 spreadsheets into a single `DataFrame`.\n",
    "Then, add the metadata information from the 10<sup>th</sup> spreadsheet as columns in the combined `DataFrame`.\n",
    "Make sure that the final `DataFrame` has a unique index and all the `NaN` values have been replaced by the tag `unknown`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>unknows</th>\n",
       "      <th colspan=\"4\" halign=\"left\">tissue</th>\n",
       "      <th colspan=\"4\" halign=\"left\">stool</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GROUP</th>\n",
       "      <th>EXTRACTION CONTROL</th>\n",
       "      <th>NEC 1</th>\n",
       "      <th>Control 1</th>\n",
       "      <th>NEC 2</th>\n",
       "      <th>Control 2</th>\n",
       "      <th>NEC 1</th>\n",
       "      <th>Control 1</th>\n",
       "      <th>NEC 2</th>\n",
       "      <th>Control 2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BARCODE</th>\n",
       "      <th>MID1</th>\n",
       "      <th>MID2</th>\n",
       "      <th>MID3</th>\n",
       "      <th>MID4</th>\n",
       "      <th>MID5</th>\n",
       "      <th>MID6</th>\n",
       "      <th>MID7</th>\n",
       "      <th>MID8</th>\n",
       "      <th>MID9</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Ignisphaera</th>\n",
       "      <td>7</td>\n",
       "      <td>23</td>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "      <td>28</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>unknows</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Pyrodictiaceae Pyrolobus</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>unknows</td>\n",
       "      <td>unknows</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>unknows</td>\n",
       "      <td>unknows</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Sulfolobales Sulfolobaceae Stygiolobus</th>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "      <td>unknows</td>\n",
       "      <td>14</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Thermoproteales Thermofilaceae Thermofilum</th>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>unknows</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>unknows</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanocellales Methanocellaceae Methanocella</th>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>12</td>\n",
       "      <td>18</td>\n",
       "      <td>unknows</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "SAMPLE                                                        unknows tissue  \\\n",
       "GROUP                                              EXTRACTION CONTROL  NEC 1   \n",
       "BARCODE                                                          MID1   MID2   \n",
       "name                                                                           \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...                  7     23   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...                  2      2   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...                  3     10   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...                  3      9   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...                  7      9   \n",
       "\n",
       "SAMPLE                                                                 \\\n",
       "GROUP                                              Control 1    NEC 2   \n",
       "BARCODE                                                 MID3     MID4   \n",
       "name                                                                    \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...        14        2   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...   unknows  unknows   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...         4  unknows   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...         5  unknows   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...         7        1   \n",
       "\n",
       "SAMPLE                                                       stool            \\\n",
       "GROUP                                              Control 2 NEC 1 Control 1   \n",
       "BARCODE                                                 MID5  MID6      MID7   \n",
       "name                                                                           \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...        28     7         8   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...         3     2         1   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...        14     5         5   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...        10     4         5   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...        17    12        18   \n",
       "\n",
       "SAMPLE                                                                 \n",
       "GROUP                                                 NEC 2 Control 2  \n",
       "BARCODE                                                MID8      MID9  \n",
       "name                                                                   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknows        16  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknows   unknows  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...        1         6  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  unknows         5  \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknows        14  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import glob\n",
    "from os.path import basename\n",
    "import os \n",
    "\n",
    "data = pd.DataFrame(columns =['name']) \n",
    "\n",
    "# For each  MID files\n",
    "for excel in  glob.glob(DATA_FOLDER+\"microbiome/*.xls\"): \n",
    "    if \"meta\" not in excel : \n",
    "        #get name of file/name of collumn of data\n",
    "        mid =os.path.splitext(basename(excel))[0]\n",
    "        # read without header\n",
    "        subdata  = pd.read_excel(excel,header=None)\n",
    "        # rename data with mid name\n",
    "        subdata.columns = [\"name\",mid]\n",
    "        # Test if bacteril name is unique, for integirty \n",
    "        assert ( any(subdata['name'].duplicated()) == False)\n",
    "        # we merge the Dataframe with the name for common key, \"outer\" add the new bacterie\n",
    "        data = data.merge(subdata,on='name', how='outer')\n",
    "        # change na by unknowd\n",
    "        data.fillna(\"unknows\", inplace = True)\n",
    "        \n",
    "\n",
    "# set name to index, do no crash , name are unique thanks to assert\n",
    "data = data.set_index('name')\n",
    "# sort collumn name\n",
    "data = data[[cn for cn in  sorted(data.columns) ]]\n",
    "\n",
    "meta  = pd.read_excel((DATA_FOLDER+\"microbiome/metadata.xls\"))\n",
    "meta.fillna(\"unknows\", inplace = True)\n",
    "# rename collone with data form metadata\n",
    "data.columns = [i for i in reversed(meta.transpose().values.tolist())]\n",
    "# name colonum with data from metadata\n",
    "data.columns.names = [i for i in reversed(meta.transpose().index.values.tolist())]\n",
    "\n",
    "data.head()    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
